draft-ietf-httpbis-priority.md

```diff
@@ -185,136 +185,72 @@ In both cases, the set of priority parameters is encoded as a Structured Headers
-delivered to the client, but one that does not block other responses from being
-used.
-
-For example, when a user using a web browser navigates to a new HTML document,
-the request for that HTML is given the default urgency.  When that HTML document
-uses a custom font, the request for that custom font SHOULD also be given the
-default urgency.  This is because the availability of the custom font is likely
-a precondition for the user to use that portion of the HTML document, which is
-to be rendered by that font.
-
-### supplementary
-
-The supplementary urgencies (values 2 to 6) indicate a response that is helpful
-to the client using a composition of responses, even though the response itself
-is not mandatory for using those responses.
+A client that fetches a document that likely consists of multiple HTTP resources
+(e.g., HTML) SHOULD assign the default urgency level to the main resource.  This
+convention provides the servers the capability of refining the urgency using
+knowledge specific to the web-site (see {{merging}}).
 
-For example, inline images (i.e., images being fetched and displayed as part of
-the document) are visually important elements of an HTML document.  As such,
-users will typically not be prevented from using the document, at least to some
-degree, before any or all of these images are loaded. Display of those images
-are thus considered to be an improvement for visual clients rather than a
-prerequisite for all user agents.  Therefore, such images will be given the
-supplementary urgency.
-
-Values between 2 and 6 are used to represent this urgency, to provide
-flexibility to the endpoints for giving some responses more or less precedence
-than others that belong to the supplementary group. {{merging}} explains how
-these values might be used.
-
-Clients SHOULD NOT use values 2 and 6.  Servers MAY use these values to
-prioritize a response above or below other supplementary responses.
-
-Clients MAY use values 3 to indicate that a request is given relatively high
-priority, or 5 to indicate relatively low priority, within the supplementary
-urgency group.
-
-For example, an image certain to be visible at the top of the page, might be
-assigned a value of 3 instead of 4, as it will have a high visual impact for the
-user.  Conversely, an asynchronously loaded JavaScript file might be assigned an
-urgency value of 5, as it is less likely to have a visual impact.
-
-When none of the considerations above is applicable, the value of 3 SHOULD be
-used.
-
-### background
-
-The background urgency (value 7) is used for responses of which the delivery can
-be postponed without having an impact on using other responses.
-
-As an example, the download of a large file in a web browser would be assigned
-the background urgency so it would not impact further page loads on the same
-connection.
+The lowest urgency level (7) is reserved for background tasks such as delivery
```

Comment:
It might be worth specifying which urgency level should be used for server push streams. I think the "optimal" solution is to default to 7, then rely on the client to reprioritize the pushed stream when it matches an actual requests. Assuming no bufferbloat, this ensures that the server does not waste any high-priority bandwidth on pushes that won't be used.

Reply:
There are non-browser use-cases of push where background might not be the best level for pushed resources, but as long as we are not too prescriptive about it I don't object to saying that background is recommended.

Reply:
I think @tombergen's suggestion is a sensible one. We can support @afrind case by applying the server signal to the push; absence of server signal means default (in this case background), presence of signal prioritizes at a different level and, if passed back to the client, is an explicit signal of the push priority. If this matches the priority the client would have reprioritized to (because it is more important) you save the client from sending a pointless reprioritization.

Reply:
IMO, the default should (can) be 7 only when all the following conditions are met:
* 1. There is no bufferbloat.
* 2. If the specification retains the definition of a reprioritization frame.
* 3. If the browser correctly uses that.

For point 1, while there would likely be no bufferbloat in HTTP/3, there will be in HTTP/2. For point 2, the discussion is ongoing in #1021, and I am not sure if we can be certain that browsers would reprioritize pushed assets even in case of it being specified.

Considering these aspects, I tend to think that the correct thing for a server to do by default would be prioritize the pushed response right after the response that initiated the push. In terms of the specification, we can either state to use the same urgency level with i=false, or the next urgency level.

Such a design would give us what we have now in HTTP/2.

In case of HTTP/3, there would be negligible negative impact, assuming that there would be no bufferbloat. This is because URLs of pushed responses and the HTML would reach clients before the pushed resources do. Clients would be possible to cancel / reprioritize the pushed resources when or before when it discovers and issues requests for the non-pushed resources that it needs.

Reply:
I think it's fine to punt this question to #1021. You are right that this becomes problematic without reprioritization. If repriotization goes away, perhaps we'll need to say something like "servers should not push resources unless they are confident they can assign a suitable priority for that resource". 

> I tend to think that the correct thing for a server to do by default would be prioritize the pushed response right after the response that initiated the push

We have observed that H2 push can be a serious footgun. It's easy to hurt performance by over-pushing, such as by pushing low priority resources or cached resources. A level 7 default reduces the potential impact of this footgun. As @LPardue said, this is just a default; the server can always do something different when it believes it knows better.

> Clients would be possible to cancel the pushed resources when or before when it discovers

It's difficult for the client to know whether or not they should cancel a pushed resource. Maybe they will need the resource soon, they just don't know yet.

> I am not sure if we can be certain that browsers would reprioritize pushed assets even in case of it being specified.

FWIW, Chrome already does this for H2. Not sure about other browsers.

Reply:
Created #1056 so that we do not miss it.